src/server.rs

```diff
@@ -469,4 +473,112 @@ pub extern "C" fn rustls_server_session_get_sni_hostname(
         }
         rustls_result::Ok
     }
-}
\ No newline at end of file
+}
+
+/// The TLS Client Hello information provided to a ClientHelloCallback function.
+/// `sni_name` is the SNI servername provided by the client (not 0 terminated) with
+/// `sni_name_len` as its length. If the client did not provide an SNI, the name
+/// length is 0.
+#[allow(non_camel_case_types)]
+#[repr(C)]
+pub struct rustls_client_hello {
+    sni_name: *const c_char,
+    sni_name_len: size_t,
+    signature_schemes: *const c_ushort,
+    signature_schemes_len: size_t,
+}
+
+/// Any context information the callback will receive when invoked.
+type ClientHelloUserData = *mut c_void;
+
+/// Called when the TLS ClientHello message was received and successfully decoded.
+type ClientHelloCallback = unsafe extern "C" fn(
+    userdata: ClientHelloUserData,
+    hello: *const rustls_client_hello,
+) -> rustls_result;
+
+struct ClientHelloResolver {
+    /// Implementation of rustls::ResolvesServerCert that passes values
+    /// from the supplied ClientHello to the callback function.
+    pub callback: ClientHelloCallback,
+    pub userdata: ClientHelloUserData,
+    pub previous_resolver: Option<Arc<dyn rustls::ResolvesServerCert>>,
+}
+
+impl ClientHelloResolver {
+    pub fn new(
+        callback: ClientHelloCallback,
+        userdata: ClientHelloUserData,
+        previous_resolver: Option<Arc<dyn rustls::ResolvesServerCert>>,
+    ) -> ClientHelloResolver {
+        ClientHelloResolver { callback, userdata, previous_resolver }
+    }
+}
+
+impl rustls::ResolvesServerCert for ClientHelloResolver {
+    fn resolve(&self, client_hello: ClientHello<'_>) -> Option<CertifiedKey> {
+        let sni_name = {
+            match client_hello.server_name() {
+                Some(c) => c.into(),
+                None => "",
+            }
+        };
+        let sigschemes = map_signature_schemes(client_hello.sigschemes());
+        let hello = rustls_client_hello {
+            sni_name: sni_name.as_ptr() as *const c_char,
+            sni_name_len: sni_name.len() as size_t,
+            signature_schemes: sigschemes.as_ptr(),
+            signature_schemes_len: sigschemes.len(),
+        };
+        let cb = self.callback;
+        let result: rustls_result = unsafe { cb(self.userdata, &hello) };
+        match result {
+            rustls_result::Ok => rustls_result::Ok,
+            r => panic!("client hello callback returned: {}", r)
+        };
+        match &self.previous_resolver {
+            Some(r) => r.resolve(client_hello),
+            None => None
+        }
+    }
+}
+
+unsafe impl marker::Sync for ClientHelloResolver {}
+unsafe impl marker::Send for ClientHelloResolver {}
```

Comment:
Send and Sync are in [`std::prelude`](https://doc.rust-lang.org/std/prelude/index.html), so you don't need to preface them with `marker::` (and it's somewhat unidiomatic to do so).

Reply:
Ack.